src/datasets/features/image.py

```diff
@@ -0,0 +1,287 @@
+            else:
+                return {"path": None, "bytes": image_to_bytes(value)}
+        else:
+            return value
+
+    def decode_example(self, value):
+        """Decode example image file into image data.
+
+        Args:
+            value (obj:`str` or :obj:`dict`): a string with the absolute image file path, a dictionary with
+                keys:
+                - path: String with absolute or relative image file path.
+                - bytes: The bytes of the image file.
+
+        Returns:
+            :obj:`PIL.Image.Image`
+        """
+        if config.PIL_AVAILABLE:
+            import PIL.Image
+        else:
+            raise ImportError("To support decoding images, please install 'Pillow'.")
+
+        if isinstance(value, str):
+            path, bytes_ = value, None
+        else:
+            path, bytes_ = value["path"], value["bytes"]
+
+        if bytes_ is None:
+            if isinstance(path, str):
+                if is_local_path(path):
+                    image = PIL.Image.open(path)
+                else:
+                    with xopen(path, "rb") as f:
+                        bytes_ = BytesIO(f.read())
+                    image = PIL.Image.open(bytes_)
+        else:
+            image = PIL.Image.open(BytesIO(bytes_))
+        return image
+
+
+def list_image_compression_formats():
+    if config.PIL_AVAILABLE:
+        import PIL.Image
+    else:
+        raise ImportError("To support encoding images, please install 'Pillow'.")
+
+    global _IMAGE_COMPRESSION_FORMATS
+    if _IMAGE_COMPRESSION_FORMATS is None:
+        PIL.Image.init()
+        _IMAGE_COMPRESSION_FORMATS = list(set(PIL.Image.OPEN.keys()) & set(PIL.Image.SAVE.keys()))
+    return _IMAGE_COMPRESSION_FORMATS
+
+
+def image_to_bytes(image: "PIL.Image.Image") -> bytes:
+    """Convert a PIL Image object to bytes using native compression if possible, otherwise use PNG compression."""
+    buffer = BytesIO()
+    format = image.format if image.format in list_image_compression_formats() else "PNG"
+    image.save(buffer, format=format)
+    return buffer.getvalue()
+
+
+def objects_to_list_of_image_dicts(objs):
+    """Encode a list of objects into a format suitable for creating an extension array of type :obj:`ImageExtensionType`."""
+    if config.PIL_AVAILABLE:
+        import PIL.Image
+    else:
+        raise ImportError("To support encoding images, please install 'Pillow'.")
+
+    if objs:
+        obj = objs[0]
+        if isinstance(obj, np.ndarray):
+            return [{"path": None, "bytes": image_to_bytes(PIL.Image.fromarray(obj.astype(np.uint8)))} for obj in objs]
+        elif isinstance(obj, PIL.Image.Image):
+            return [
+                {"path": obj.filename, "bytes": None}
+                if hasattr(obj, "filename") and obj.filename != ""
+                else {"path": None, "bytes": image_to_bytes(obj)}
+                for obj in objs
+            ]
+        else:
```

Comment:
does it also have to support lists of strings as input ?
```suggestion
        elif isinstance(obj, str):
            return [{"path": obj, "bytes": None} for obj in objs]
        else:
```

Reply:
Only if we want support returning image file paths in transforms, e.g.
```python
def transform(example):
   ...
   return image_path

dset.map(transform, features=Features("image": Image(), ...))
```
I would like to avoid that because I don't see this being used at all, and could lead to some weird bugs during decoding if the returned paths don't point to valid image files.

@nateraw @sgugger Opinions?

Reply:
You mean this ?
```python
def transform(example):
   ...
   return {"image": image_path}
```

I think it would be nice to have it, to be consistent with `from_dict`:
```python
dataset = Dataset.from_dict({"image": [image_path]}, features=Features({"image": Image()}))
```

It also sounds much easier fore the users to do this rather than
```python
def transform(example):
   ...
   return {"image": {"path": image_path, "bytes": None}}
```